src/storage/mod.rs

```diff
@@ -4256,4 +4262,484 @@ mod tests {
+        }
+    }
+
+    impl LockManager for ProxyLockMgr {
+        fn wait_for(
+            &self,
+            start_ts: TimeStamp,
+            cb: StorageCallback,
+            pr: ProcessResult,
+            lock: Lock,
+            is_first_lock: bool,
+            timeout: Option<WaitTimeout>,
+        ) {
+            self.tx
+                .send(Msg::WaitFor {
+                    start_ts,
+                    cb,
+                    pr,
+                    lock,
+                    is_first_lock,
+                    timeout,
+                })
+                .unwrap();
+        }
+
+        fn wake_up(
+            &self,
+            lock_ts: TimeStamp,
+            hashes: Vec<u64>,
+            commit_ts: TimeStamp,
+            is_pessimistic_txn: bool,
+        ) {
+            self.tx
+                .send(Msg::WakeUp {
+                    lock_ts,
+                    hashes,
+                    commit_ts,
+                    is_pessimistic_txn,
+                })
+                .unwrap();
+        }
+
+        fn has_waiter(&self) -> bool {
+            self.has_waiter.load(Ordering::Relaxed)
+        }
+    }
+
+    // Test whether `Storage` sends right wait-for-lock msgs to `LockManager`.
+    #[test]
+    fn validate_wait_for_lock_msg() {
+        let (msg_tx, msg_rx) = channel();
+        let storage = TestStorageBuilder::from_engine(TestEngineBuilder::new().build().unwrap())
+            .set_lock_mgr(ProxyLockMgr::new(msg_tx))
+            .build()
+            .unwrap();
+
+        let (k, v) = (b"k".to_vec(), b"v".to_vec());
+        let (tx, rx) = channel();
+        // Write lock-k.
+        storage
+            .sched_txn_command(
+                commands::Prewrite::with_defaults(
+                    vec![Mutation::Put((Key::from_raw(&k), v))],
+                    k.clone(),
+                    10.into(),
+                ),
+                expect_ok_callback(tx.clone(), 0),
+            )
+            .unwrap();
+        rx.recv().unwrap();
+        // No wait for msg
+        assert!(msg_rx.try_recv().is_err());
+
+        // Meet lock-k.
+        storage
+            .sched_txn_command(
+                commands::AcquirePessimisticLock::new(
+                    vec![
+                        (Key::from_raw(b"foo"), false),
+                        (Key::from_raw(&k.clone()), false),
+                    ],
+                    k.clone(),
+                    20.into(),
+                    3000,
+                    true,
+                    20.into(),
+                    Some(WaitTimeout::Millis(100)),
+                    false,
+                    21.into(),
+                    Context::default(),
+                ),
+                expect_ok_callback(tx, 0),
+            )
+            .unwrap();
+        // The transaction should be waiting for lock released so cb won't be called.
+        rx.recv_timeout(Duration::from_millis(500)).unwrap_err();
```

Comment:
Could we just use `recv` here? Or set the timeout much larger (e.g., one minute) so that we don't hit as often - timeouts are a big source of non-deterministic tests.

Reply:
No. The timeout means it shouldn't recv msg here...